File: ruby/beer-song/beer_song.rb
class BeerSong
  def lyrics
    verses(99, 0)
  end

  def verses(start, finish)
    start.downto(finish).map { |n| verse(n) }.join("\n")
  end

  def verse(count)
    "#{bottles(count).capitalize} of beer on the wall, #{bottles(count)} of beer.\n" \
      "#{what_to_do(count).capitalize}, #{bottles(count-1)} of beer on the wall.\n"
  end

  private

  def what_to_do(count)
    case count
    when 0
      "Go to the store and buy some more"
    when 1
      "Take it down and pass it around"
    else
      "Take one down and pass it around"
    end
  end

  def bottles(count)
    if count == 1
      "1 bottle"
    else
      "#{remaining(count)} bottles"
    end
  end

  def remaining(count)
    if count < 0
      "99"
    elsif count == 0
      "no more"
    else
      count
    end
  end
end

module BookKeeping
  VERSION = 2
end

File: elixir/beer-song/beer_song.exs
defmodule BeerSong do
  @doc """
  Get a single verse of the beer song
  """
  @spec verse(integer) :: String.t
  def verse(number) do
    count = number-1
    """
    #{bottles(count) |> String.capitalize} of beer on the wall, #{bottles(count)} of beer.
    #{bottle_action(count)}, #{bottles(count-1)} of beer on the wall.
    """
  end

  defp bottles(1), do: "1 bottle"
  defp bottles(count), do: "#{remaining count} bottles"

  defp bottle_action(0), do: "Go to the store and buy some more"
  defp bottle_action(1), do: "Take it down and pass it around"
  defp bottle_action(n), do: "Take one down and pass it around"

  defp remaining(0), do: "no more"
  defp remaining(n) when n < 0, do: 99
  defp remaining(n), do: n

  @doc """
  Get the entire beer song for a given range of numbers of bottles.
  """
  @spec lyrics(Range.t) :: String.t
  def lyrics, do: lyrics(100..1)
  def lyrics(range) do
    range
    |> Enum.map(fn n -> verse(n) end)
    |> Enum.join("\n")
  end
end
